src/adapter/src/catalog/open.rs

```diff
@@ -1153,6 +1297,61 @@ impl BuiltinBootstrapClusterSizes {
     }
 }
 
+#[derive(Debug, Default)]
+pub struct SystemViewDependenciesVisitor {
+    names: BTreeSet<UnresolvedItemName>,
+}
+
+// Visitor that records the named dependencies for a statement.
+impl<'ast> Visit<'ast, Raw> for SystemViewDependenciesVisitor {
+    fn visit_item_name(&mut self, item_name: &'ast <Raw as AstInfo>::ItemName) {
+        let RawItemName::Name(name) = item_name else {
+            return;
+        };
+        // Dependent views must have only a schema specified.
+        if name.0.len() != 2 {
+            return;
+        }
+        self.names.insert(name.clone());
+    }
+
+    fn visit_expr(&mut self, node: &'ast Expr<Raw>) {
+        // Casts to reg* types involve reading from views, so any view with these casts must also
+        // depend on those views. In typeconv.rs see STRING_REG_CAST_TEMPLATE for raw SQL. If the
+        // view dependencies there change, this must also be updatedj.
+        if let Expr::Cast {
+            data_type:
+                RawDataType::Other {
+                    name: RawItemName::Name(name),
+                    ..
+                },
+            ..
+        } = node
+        {
+            if name.0.len() == 2
+                && name.0[0].as_str().to_ascii_lowercase() == "pg_catalog"
+                && name.0[1].as_str().to_ascii_lowercase().starts_with("reg")
+            {
+                self.names.insert(UnresolvedItemName(vec![
+                    ident!("mz_catalog"),
+                    ident!("mz_objects"),
+                ]));
+            }
+        }
+        visit_expr(self, node);
+    }
+}
+
+/// Returns the set of named dependencies for a statement.
+pub fn visit_system_view_dependency_names<'ast, N>(node: &'ast N) -> BTreeSet<UnresolvedItemName>
+where
+    N: VisitNode<'ast, Raw> + 'ast,
+{
+    let mut visitor = SystemViewDependenciesVisitor::default();
+    node.visit(&mut visitor);
+    visitor.names
+}
```

Comment:
I'm worried that this might be too brittle for collecting dependencies. It seems too easy for someone to implement a new cast/function using a catalog view without updating this or create a new builtin without fully qualified names. Have you explored using the approach we currently use with items to collect dependencies? As in, try parsing all (or some large batch of) views in parallel and learn the dependencies of the view based on the error.

Reply:
Yes, this was one approach that I implemented. Although it worked, I thought it was similarly hacky. It needs to call `resolve` possibly multiple times for each object and sniff out errors. The solution in the current commit, while brittle and subject to the problems you describe, does have the nice property that it is single-pass over each object.

It's very silly that we have full information available at compile time to determine the dependency order, but we use the worst possible way of "try as many times as needed and sniff out errors" instead of doing something more principled.

[note: this paragraph led to an improved idea for runtime dependency finding and I no longer endorse a build.rs step]
If we were to brainstorm the correct way to solve the problem "for a known set of sql functions, types, and objects, compute the dependency graph such that it is efficiently available at runtime", what would we do? My first thought is a build.rs step that could compute this and provide it as some data structure with immediate dependency lookup for any builtin. It'd be swell if it had some implementation that could do a single pass over each object.

----

> It seems too easy for someone to implement a new cast/function using a catalog view without updating this or create a new builtin without fully qualified names.

This problem will be discovered by the person doing the programming and immediately discovered when they start up the catalog because it'll panic. I think that's totally fine because it will be obvious that their new change contributed to that. We should however make this the most pleasant experience possible. What if the error message when this happens is so helpful that it's very easy to know what to do? No one will have to remember it or lose time trying to figure out what went wrong. And because it happens deterministically at startup, it's not even a thing that could be missed by tests or code reviews.

Reply:
That all makes sense to me, I think the error message idea is good to and your current error message looks good to me.